src/kernel/src/fs/mod.rs

```diff
@@ -409,11 +440,106 @@ impl Fs {
 
         // TODO: It seems like the initial ucred of the process is either root or has PRIV_VFS_ADMIN
         // privilege.
-        self.revoke(vn, td.deref())?;
+        self.revoke(vn, &td)?;
+
+        Ok(SysOut::ZERO)
+    }
+
+    fn sys_stat(self: &Arc<Self>, i: &SysIn) -> Result<SysOut, SysErr> {
+        let path = unsafe { i.args[0].to_path() }?.unwrap();
+        let stat_out: *mut Stat = i.args[1].into();
+
+        let td = VThread::current().unwrap();
+
+        let stat = self.stat(path, &td)?;
+
+        unsafe {
+            *stat_out = stat;
+        }
+
+        Ok(SysOut::ZERO)
+    }
+
+    /// This function is inlined on the PS4, but corresponds to `kern_stat` in FreeBSD.
+    fn stat(self: &Arc<Self>, path: impl AsRef<VPath>, td: &VThread) -> Result<Stat, StatError> {
+        self.statat(AtFlags::empty(), At::Cwd, path, td)
+    }
+
+    fn sys_fstat(self: &Arc<Self>, i: &SysIn) -> Result<SysOut, SysErr> {
+        let fd: i32 = i.args[0].try_into().unwrap();
+        let stat_out: *mut Stat = i.args[1].into();
+
+        let td = VThread::current().unwrap();
+
+        let stat = self.fstat(fd, &td)?;
+
+        unsafe {
+            *stat_out = stat;
+        }
+
+        Ok(SysOut::ZERO)
+    }
+
+    /// See `kern_fstat` on the PS4 for a reference.
+    #[allow(unused_variables)] // Remove this when it is being implemented
+    fn fstat(self: &Arc<Self>, fd: i32, td: &VThread) -> Result<Stat, StatError> {
+        todo!()
+    }
+
+    fn sys_lstat(self: &Arc<Self>, i: &SysIn) -> Result<SysOut, SysErr> {
+        let path = unsafe { i.args[0].to_path() }?.unwrap();
+        let stat_out: *mut Stat = i.args[1].into();
+
+        let td = VThread::current().unwrap();
+
+        td.priv_check(Privilege::SCE683)?;
+
+        let stat = self.lstat(path, &td)?;
+
+        unsafe {
+            *stat_out = stat;
+        }
+
+        Ok(SysOut::ZERO)
+    }
+
+    /// See `kern_lstat` in FreeBSD for a reference. (This function is inlined on the PS4)
+    fn lstat(self: &Arc<Self>, path: impl AsRef<VPath>, td: &VThread) -> Result<Stat, StatError> {
+        self.statat(AtFlags::SYMLINK_NOFOLLOW, At::Cwd, path, td)
+    }
+
+    fn sys_fstatat(self: &Arc<Self>, i: &SysIn) -> Result<SysOut, SysErr> {
+        let dirfd: i32 = i.args[0].try_into().unwrap();
+        let path = unsafe { i.args[1].to_path() }?.unwrap();
+        let stat_out: *mut Stat = i.args[2].into();
+        let flags: AtFlags = i.args[3].try_into().unwrap();
+
+        let td = VThread::current().unwrap();
+
+        td.priv_check(Privilege::SCE683)?;
```

Comment:
@red-prig Any idea what this privilege is?

Reply:
@SuchAFuriousDeath I don’t know what this priv constant is called and I don’t think it will be possible to find out, so we can judge it only by context

Reply:
That's what I thought. thanks anyway.